File: ecommerce-frontend/src/api/coupon.ts
import apiClient from './client';

export interface CouponValidateRequest {
  code: string;
  order_amount: number;
}

export interface CouponValidateResponse {
  valid: boolean;
  code: string;
  discount_type: string;
  discount_value: number;
  discount_amount: number;
  final_amount: number;
  message: string;
}

export interface CouponAvailableItem {
  coupon_id: number;
  code: string;
  discount_type: string;
  discount_value: number;
  min_order_amount: number;
  discount_amount: number;
  final_amount: number;
  applicable: boolean;
  reason: string;
  expires_at: string | null;
  is_claimed: boolean;
}

export interface CouponAdminItem {
  id: number;
  code: string;
  discount_type: 'percentage' | 'fixed';
  discount_value: number;
  min_order_amount: number;
  is_active: boolean;
  usage_limit: number;
  used_count: number;
  expires_at: string | null;
  created_at: string;
}

export interface UserCouponItem {
  id: number;
  status: 'claimed' | 'used';
  claimed_at: string;
  used_at: string | null;
  coupon: CouponAdminItem;
}

export const couponApi = {
  validate: async (data: CouponValidateRequest): Promise<CouponValidateResponse> => {
    const response = await apiClient.post('/coupons/validate', data);
    return response.data;
  },

  getAvailable: async (orderAmount: number, claimedOnly = false): Promise<CouponAvailableItem[]> => {
    const response = await apiClient.get('/coupons/available', {
      params: { order_amount: orderAmount, claimed_only: claimedOnly }
    });
    return response.data;
  },

  claim: async (couponId: number): Promise<UserCouponItem> => {
    const response = await apiClient.post(`/coupons/claim/${couponId}`);
    return response.data;
  },

  mine: async (status?: string): Promise<UserCouponItem[]> => {
    const response = await apiClient.get('/coupons/mine', { params: status ? { status } : undefined });
    return response.data;
  }
};

export const couponAdminApi = {
  list: async (): Promise<CouponAdminItem[]> => {
    const response = await apiClient.get('/admin/coupons/');
    return response.data;
  },

  create: async (data: Omit<CouponAdminItem, 'id' | 'used_count' | 'created_at'>): Promise<CouponAdminItem> => {
    const response = await apiClient.post('/admin/coupons/', data);
    return response.data;
  },

  update: async (id: number, data: Partial<Omit<CouponAdminItem, 'id' | 'used_count' | 'created_at'>>): Promise<CouponAdminItem> => {
    const response = await apiClient.put(`/admin/coupons/${id}`, data);
    return response.data;
  },

  delete: async (id: number): Promise<void> => {
    await apiClient.delete(`/admin/coupons/${id}`);
  }
};

File: ecommerce-backend/app/schemas/coupon.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class CouponCreate(BaseModel):
    code: str = Field(..., min_length=1, max_length=50)
    discount_type: str = Field(default="fixed", pattern="^(percentage|fixed)$")
    discount_value: float = Field(..., gt=0)
    min_order_amount: float = Field(default=0, ge=0)
    is_active: bool = True
    usage_limit: int = Field(default=0, ge=0)
    expires_at: Optional[datetime] = None


class CouponUpdate(BaseModel):
    code: Optional[str] = Field(None, min_length=1, max_length=50)
    discount_type: Optional[str] = Field(None, pattern="^(percentage|fixed)$")
    discount_value: Optional[float] = Field(None, gt=0)
    min_order_amount: Optional[float] = Field(None, ge=0)
    is_active: Optional[bool] = None
    usage_limit: Optional[int] = Field(None, ge=0)
    expires_at: Optional[datetime] = None


class CouponOut(BaseModel):
    id: int
    code: str
    discount_type: str
    discount_value: float
    min_order_amount: float
    is_active: bool
    usage_limit: int
    used_count: int
    expires_at: Optional[datetime] = None
    created_at: datetime

    class Config:
        from_attributes = True


class CouponValidateRequest(BaseModel):
    code: str = Field(..., min_length=1)
    order_amount: float = Field(..., ge=0)


class CouponValidateResponse(BaseModel):
    valid: bool
    code: str
    discount_type: str
    discount_value: float
    discount_amount: float = 0
    final_amount: float = 0
    message: str = ""


class CouponAvailableItem(BaseModel):
    coupon_id: int
    code: str
    discount_type: str
    discount_value: float
    min_order_amount: float
    discount_amount: float
    final_amount: float
    applicable: bool
    reason: str = ""
    expires_at: str | None = None
    is_claimed: bool = False


class UserCouponOut(BaseModel):
    id: int
    status: str
    claimed_at: datetime
    used_at: Optional[datetime] = None
    coupon: CouponOut

    class Config:
        from_attributes = True
